### ISSUE TITLE
Apparent "memory leak"/performance degradation, seemingly caused by CoFHCore

### ISSUE THREAD

**Issue by dvdmandt** [GTNewHorizons/GT-New-Horizons-Modpack#6964]
#### Which modpack version are you using?
2.0.9.0QF4
#
#### If in multiplayer; On which server does this happen? 
My own server
#
#### What did you try to do, and what did you expect to happen?
Let the server run for a few days without performance degrading to a crawl.
#
#### What happened instead? (Attach screenshots if needed)
Performance degraded to unplayable. 

Me and @Glease have been looking into it. It seems whenever a TileEntity is added to the world, by placement or chunk loading (?), it is added to a LinkedHashList which is effectively never cleared. Since loading chunks will create new instances of the entities, the list will continue to grow and keep more and more copies of the TileEntities in memory, including any NBT data etc they may contain. This applies to machines etc, but also GT Ores it seems.

This seems to be the code responsible: https://github.com/CoFH/CoFHCore/blob/4be193afe63bcf108c5ad5abd34da34028434c65/src/main/java/cofh/asm/ASMCore.java#L1258-L1454

This probably affects all single and multiplayer worlds, but some setups may make it worse. In my case, I had a Thaumcraft mirror pair where one end was loaded with an anchor and the other one unloaded on Mars. My guess is that the loaded mirror in my base force loaded the Mars one every now and then, adding any nearby ore blocks to the list, resulting in millions of entries after a while. There were also thousands of copies of the mars end of the mirror left in memory, as keys in the cofh_recentTiles table. But as far as I can tell, any exploration, or movement in unloaded chunks will eventually trigger this issue. Anything causing repeated loading of chunks with TileEntities in them.

#
#### What do you suggest instead/what changes do you propose?
Fix it through code, ASM, Mixins or similar, or, worst case, get rid of the mod completely if we can and don't need it.

**Comment by dvdmandt** [comment]
![image](https://user-images.githubusercontent.com/11985732/99778257-bb333380-2b13-11eb-9f32-a8130ca5371e.png)

This was what my VisualVM reported after 44 hours of uptime. My estimation is that 95%+ of those GT Ores and LinkedHashList$Entry instances were caused by this issue, and since it spawns copies of all NBT data of all blocks loaded multiple times, likely a large part of those strings and HashMap$Node's as well.

**Comment by Glease** [comment]
https://github.com/Glease/CoFHCoreFix
 a temporary fix. It's currently under testing with @dvdmandt server. We will see how this can improve the RAM usage.

**Comment by dvdmandt** [comment]
My server has been running with this fix for a few days now and no issues have popped up so far. It seems to do its intended job, but unfortunately the issue that triggered this for me also triggers #6976 which in my scenario causes nearly identical "leaks", so it's hard to verify the efficiency conclusively at this point. Given what I've seen so far, I would expect the proposed fix to have a significant impact on performance over time for any servers with medium to high population and activity.

**Comment by fluffle** [comment]
I dumped a heap from my server and can confirm the IdentityLinkedHashList mentioned upthread is completely empty for me, so I suspect it is not this bug causing duped tile entities. OTOH, I have ... 23,694 TileEntityReactorChamberElectric and 3,945 TileEntityNuclearReactorElectric in this heap, and I know for sure that I've only built 2 6-chamber reactors, so *something* is definitely leaking. I will see if I can find other related bugs.
